### PR TITLE
C-API Metadata

### PR DESCRIPTION
**Pull request by atobiszei**
* ModelInstance metadata support
* DAG metadata support
* dynamic shapes support
* ModelAPI adapter using C-API metadata

JIRA:CVS-107509

### CODE REVIEW

File: src/capi_frontend/capi.cpp [hunk 1]
```diff
@@ -683,6 +709,141 @@ OVMS_Status* OVMS_Inference(OVMS_Server* serverPtr, OVMS_InferenceRequest* reque
     return nullptr;
 }
 
+OVMS_Status* OVMS_ServableMetadataGet(OVMS_Server* serverPtr, const char* servableName, int64_t servableVersion, OVMS_ServableMetadata** servableMetadata) {
```
Comment (dkalinowski) [inline src/capi_frontend/capi.cpp:712]:
```suggestion
OVMS_Status* OVMS_GetServableMetadata(OVMS_Server* serverPtr, const char* servableName, int64_t servableVersion, OVMS_ServableMetadata** servableMetadata) {
```

Comment (dkalinowski) [reply src/capi_frontend/capi.cpp:712]: Or just ServableMetadata

File: src/capi_frontend/capi.cpp [hunk 2]
```diff
@@ -683,6 +709,141 @@ OVMS_Status* OVMS_Inference(OVMS_Server* serverPtr, OVMS_InferenceRequest* reque
...
+        return reinterpret_cast<OVMS_Status*>(new Status(StatusCode::NONEXISTENT_METADATA));
+    }
+    if (count == nullptr) {
+        return reinterpret_cast<OVMS_Status*>(new Status(StatusCode::NONEXISTENT_NUMBER));
+    }
+    ovms::ServableMetadata* metadata = reinterpret_cast<ovms::ServableMetadata*>(servableMetadata);
+    *count = metadata->getOutputsInfo().size();
+    return nullptr;
+}
+OVMS_Status* OVMS_ServableMetadataGetInput(OVMS_ServableMetadata* servableMetadata, uint32_t id, const char** name, OVMS_DataType* datatype, size_t* dimCount, int64_t** shapeMin, int64_t** shapeMax) {
```
Comment (dkalinowski) [inline src/capi_frontend/capi.cpp:775]:
```suggestion
OVMS_Status* OVMS_ServableMetadataGetInput(OVMS_ServableMetadata* servableMetadata, uint32_t id, const char** name, OVMS_DataType* datatype, size_t* dimCount, int64_t** shapeMinArr, int64_t** shapeMaxArr) {
```
What about adding such suffix to indicate its an array?

File: src/mediapipe_calculators/modelapiovmsadapter.cc [hunk 3]
```diff
@@ -139,15 +141,67 @@ InferenceOutput OVMSInferenceAdapter::infer(const InferenceInput& input) {
...
+        throw std::runtime_error(std::string("Adapter could not find input:") + inputName);  // TODO error handling
+    }
+
+    ov::Shape ovShape;
+    const auto& [minBorder, maxBorder] = it->second;
+    ovShape.reserve(minBorder.size());
+    for (const auto& d : minBorder) {
+        ovShape.emplace_back(d);
+    }
+    ov::Shape ovShape2;
```
Comment (dkalinowski) [inline src/mediapipe_calculators/modelapiovmsadapter.cc:190]: add todo for partial

File: src/ovms.h [hunk 4]
```diff
@@ -434,11 +436,75 @@ void OVMS_InferenceResponseDelete(OVMS_InferenceResponse* response);
...
 //
+// \param server The server object
 // \param request The request object
 // \param response The respons object. In case of success, caller takes the ownership of the response
 // \return OVMS_Status object in case of failure
 OVMS_Status* OVMS_Inference(OVMS_Server* server, OVMS_InferenceRequest* request, OVMS_InferenceResponse** response);
 
+// Get OVMS_ServableMetadata object
+//
+// Returned object needs to be deleted after use with OVMS_ServableMetadataDelete
```
Comment (dkalinowski) [inline src/ovms.h:447]: Add info what does the call do

**Review by rasapala — APPROVED** [review]
LGTM

File: src/servablemetadata.cpp [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+ServableMetadata::ServableMetadata(const std::string& name,
+    model_version_t version,
+    tensor_map_t&& inputsInfo,
+    tensor_map_t&& outputsInfo) :
+    name(name),
+    version(version),
+    inputsInfo(inputsInfo),
+    outputsInfo(outputsInfo) {
+    for (auto& [key, tensorInfo] : this->inputsInfo) {
```
Comment (dkalinowski) [inline src/servablemetadata.cpp:53]: Body of this constructor is exact copy of the one before - what is the benefit of having two? Is the benefit really greater than code duplication? Metadata call is not on critical path anyway.

File: src/servablemetadata.hpp [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+class ServableMetadata {
+    const std::string name;
+    const model_version_t version;
+    tensor_map_t inputsInfo;
+    tensor_map_t outputsInfo;
+    capi_tensor_shapes_map_t inDimMin;
+    capi_tensor_shapes_map_t inDimMax;
+    capi_tensor_shapes_map_t outDimMin;
+    capi_tensor_shapes_map_t outDimMax;
+    // 2nd for outputs
```
Comment (dkalinowski) [inline src/servablemetadata.hpp:38]: ?

File: src/capi_frontend/capi.cpp [hunk 7]
```diff
@@ -683,6 +709,141 @@ OVMS_Status* OVMS_Inference(OVMS_Server* serverPtr, OVMS_InferenceRequest* reque
...
+    if (servableMetadata == nullptr) {
+        return reinterpret_cast<OVMS_Status*>(new Status(StatusCode::NONEXISTENT_METADATA));
+    }
+    if (name == nullptr) {
+        return reinterpret_cast<OVMS_Status*>(new Status(StatusCode::NONEXISTENT_STRING));
+    }
+    if (datatype == nullptr) {
+        return reinterpret_cast<OVMS_Status*>(new Status(StatusCode::NONEXISTENT_NUMBER));
```
[the remark this answers is not in the document]
Comment (atobiszei) [reply src/capi_frontend/capi.cpp:783]: Following convention here. We decided to not create separate status code for each input. Still having at least some differentiation will make debug easier.

File: src/capi_frontend/capi.cpp [hunk 8]
```diff
@@ -683,6 +709,141 @@ OVMS_Status* OVMS_Inference(OVMS_Server* serverPtr, OVMS_InferenceRequest* reque
...
+    }
+    if (shapeMin == nullptr) {
+        return reinterpret_cast<OVMS_Status*>(new Status(StatusCode::NONEXISTENT_TABLE));
+    }
+    if (shapeMax == nullptr) {
+        return reinterpret_cast<OVMS_Status*>(new Status(StatusCode::NONEXISTENT_TABLE));
+    }
+    ovms::ServableMetadata* metadata = reinterpret_cast<ovms::ServableMetadata*>(servableMetadata);
+    if (id >= metadata->getInputsInfo().size()) {
```
[the remark this answers is not in the document]
Comment (atobiszei) [reply src/capi_frontend/capi.cpp:827]: Why? We validate here that someone is not passing id higher than input/output count. We could segfault later on without this check.

File: src/servablemetadata.hpp (hunk 6 above)
Comment (atobiszei) [reply src/servablemetadata.hpp:38]: To remove

File: src/capi_frontend/capi.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (dkalinowski) [reply src/capi_frontend/capi.cpp:827]: Why do we need to validate against `getInputInfo` if this is `OVMS_ServableMetadataGetOutput` call?

File: src/capi_frontend/capi.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (atobiszei) [reply src/capi_frontend/capi.cpp:783]: TODO in next tasks to return 1 status code with details

**Pull request merged**